DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp

```diff
@@ -0,0 +1,200 @@
+                              ac_fixed<20, 10, true, AC_RND, AC_SAT> &ret) {
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> inp_buffer(&a, 1);
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor in_acc{inp_buffer, h, read_only};
+    accessor out_acc{ret_buffer, h, write_only};
+
+    h.single_task<ConstructFromACFixed>([=] {
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> t(in_acc[0]);
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> some_offset(0.5f);
+      out_acc[0] = t + some_offset;
+    });
+  });
+}
+
+void TestCalculateWithFloat(queue &q, const float &x, float &ret) {
+  buffer<float, 1> inp_buffer(&x, 1);
+  buffer<float, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithFloat>([=] {
+      float sin_ret = sinf(x[0]);
+      float cos_ret = cosf(x[0]);
+      res[0] = sqrtf(sin_ret * sin_ret + cos_ret * cos_ret);
+    });
+  });
+}
+
+// clang-format off
+// Please refer to ac_fixed_math.hpp header file for fixed point math
+// functions' type deduction rule. In this case, following those rules:
+// I, W, S are input type template parameter (ac_fixed<W, I, S>)
+// rI, rW, rS are output type template parameter (ac_fixed<rW, rI, rS>)
+//* Function Name         Type Propagation Rule
+//* sqrt_fixed             rI = I, rW = W, rS = S
+//* sin_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+//* cos_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+// clang-format on
+void TestCalculateWithACFixed(queue &q, const fixed_10_3_t &x,
+                              fixed_9_2_t &ret) {
+  buffer<fixed_10_3_t, 1> inp_buffer(&x, 1);
+  buffer<fixed_9_2_t, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithACFixed>([=] {
+      fixed_9_2_t sin_ret = sin_fixed(x[0]);
+      fixed_9_2_t cos_ret = cos_fixed(x[0]);
+      // The RHS expression evaluates to a larger `ac_fixed`, which gets
+      // truncated to fit in res[0].
+      res[0] = sqrt_fixed(sin_ret * sin_ret + cos_ret * cos_ret);
+    });
+  });
+}
+
+constexpr int kSize = 5;
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector selector;
+#else
+  ext::intel::fpga_selector selector;
+#endif
+
+  try {
+    // Create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    // I. Constructing `ac_fixed` Numbers
+    // Rounding mode: use `AC_RND` to round towards plus infinity
+    // Overflow mode: use `AC_SAT` to saturate to max and min
+    ac_fixed<20, 10, true, AC_RND, AC_SAT> a;
```

Comment:
What are all these `AC_RND` and `AC_SAT`? Where do they come from? Some macros defined somewhere?

Reply:
The problem is to understand where they come from.
Is `ac_fixed` defined in `::`? Same for `AC_RND` and `AC_SAT` or are they macros (argh!)? It does not fit well in a good looking C++ program.

Reply:
`ac_fixed` is defined in `::`.
Looks like`AC_RND` and `AC_SAT` are enums in `::` (from ac_int.hpp around line 1234):

```cpp
enum ac_q_mode {
  AC_TRN,
  AC_RND,
  AC_TRN_ZERO,
  AC_RND_ZERO,
  AC_RND_INF,
  AC_RND_MIN_INF,
  AC_RND_CONV,
  AC_RND_CONV_ODD
};
```

A decent change may be moving from an `enum` to an `enum class`, but these files are from Mentor Graphics, so we many need to take up any changes with them. For now, it may be worthwhile adding to the comment where these values are coming from.

Reply:
Yes, and having `ac::trn` instead of `AC_TRN`, but this is another battle to have. :-)

Reply:
agreed :|